transaction/core/src/ring_signature/rct_bulletproofs.rs

```diff
@@ -170,8 +251,52 @@ impl SignatureRctBulletproofs {
             let range_proof = RangeProof::from_bytes(&self.range_proof_bytes)
                 .map_err(|_e| Error::RangeProofError)?;
 
-            check_range_proofs(&range_proof, &commitments, &generator, rng)
+            check_range_proofs(&range_proof, &commitments, generator, rng)
                 .map_err(|_e| Error::RangeProofError)?;
+        } else {
+            // When mixed transactions are supported, self.range_proofs should contain
+            // a range proof correspond to each token id used in the transaction, in sorted
+            // order.
+            if self.range_proof_bytes.is_empty() {
+                return Err(Error::UnexpectedRangeProof);
+            }
+            if token_ids.len() != self.range_proofs.len() {
+                return Err(Error::MissingRangeProofs(
+                    token_ids.len(),
+                    self.range_proofs.len(),
+                ));
+            }
+
+            // For each used token id, and range proof, we have to pick out the matching
+            // outputs and pseudo outputs and verify the range proof.
+            for (token_id, range_proof) in token_ids.iter().zip(self.range_proofs.iter()) {
+                let generator = generator_cache.get(*token_id);
+
+                let commitments: Vec<CompressedRistretto> = self
+                    .pseudo_output_commitments
+                    .iter()
+                    .zip(self.pseudo_output_token_ids.iter())
+                    .chain(output_commitments.iter().zip(self.output_token_ids.iter()))
+                    .filter_map(|(compressed_commitment, this_token_id)| {
+                        if token_id == this_token_id {
+                            Some(compressed_commitment.point)
+                        } else {
+                            None
+                        }
+                    })
+                    .collect();
+
+                assert!(
+                    !commitments.is_empty(),
+                    "logic error when accumulating commitments for token id"
+                );
+
+                let range_proof = RangeProof::from_bytes(range_proof)
+                    .map_err(|_e| Error::RangeProofDeserializationError)?;
+
+                check_range_proofs(&range_proof, &commitments, generator, rng)
+                    .map_err(|_e| Error::RangeProofError)?;
```

Comment:
Can we include some info about the underlying error in the returned error?

Reply:
Yea let's do it

Reply:
I had to stringify the range proof error because it doesn't implement traits that `TxValidationError` requires, and I don't want to break everything